packages/amplify-gen2-codegen/src/auth/source_builder.ts

```diff
@@ -114,6 +135,131 @@ function createLogInWithPropertyAssignment(logInDefinition: LoginOptions = {}) {
     const emailDefinitionObject = factory.createObjectLiteralExpression(emailDefinitionAssignments, true);
     assignments.push(factory.createPropertyAssignment(factory.createIdentifier('email'), emailDefinitionObject));
   }
+  if (logInDefinition.googleLogin) {
+    assignments.push(
+      factory.createPropertyAssignment(
+        factory.createIdentifier('externalProviders'),
+        factory.createObjectLiteralExpression([
+          factory.createPropertyAssignment(
+            factory.createIdentifier('google'),
+            factory.createObjectLiteralExpression([
+              factory.createPropertyAssignment(
+                factory.createIdentifier('clientId'),
+                factory.createCallExpression(secretIdentifier, undefined, [factory.createStringLiteral(googleClientID)]),
+              ),
+              factory.createPropertyAssignment(
+                factory.createIdentifier('clientSecret'),
+                factory.createCallExpression(secretIdentifier, undefined, [factory.createStringLiteral(googleClientSecret)]),
+              ),
+            ]),
+          ),
+          factory.createPropertyAssignment(
+            factory.createIdentifier('callbackUrls'),
+            factory.createArrayLiteralExpression(logInDefinition.callbackURLs?.map((url) => factory.createStringLiteral(url))),
+          ),
+          factory.createPropertyAssignment(
+            factory.createIdentifier('logoutUrls'),
+            factory.createArrayLiteralExpression(logInDefinition.logoutURLs?.map((url) => factory.createStringLiteral(url))),
+          ),
+        ]),
+      ),
+    );
+  } else if (logInDefinition.appleLogin) {
+    assignments.push(
+      factory.createPropertyAssignment(
+        factory.createIdentifier('externalProviders'),
+        factory.createObjectLiteralExpression([
+          factory.createPropertyAssignment(
+            factory.createIdentifier('signInWithApple'),
+            factory.createObjectLiteralExpression([
```

Comment:
since all providers have similar code generation, can we extract this to a common function and passing the necessary parameters?

Reply:
Done [808de33](https://github.com/aws-amplify/amplify-cli/pull/13882/commits/808de33f6a38c9b2735c1f4bbfd61b6dd39522b3)